Avro metrics support: track metrics in Avro value writers

yyanyy opened the pull request:
This change is a smaller PR broken down from #1935. 

This change adds field id to constructors of Avro primitive value writers, and make these writers to track stats such as value count, min and max, and expose a `metrics` method that could be called to collect `FieldMetrics`. However nothing is calling these method yet. 
This change doesn't have any test, and tests will be included in the next PR when end to end integration is set up. 

Please note: regarding change to the signature of `FieldMetrics`, the alternative would be to keep `ByteBuffer` as the return value for lower/upper bound of `FieldMetrics` and ingest each field's metrics mode to each leaf value writer during construction, so that when collecting metrics from these writers, truncation and conversion to byte buffer could happen. I think it's doable but it would touch a lot of methods' signatures, including adding metric mode to the constructor of every leaf writer, and adding metrics config to every datum writer (e.g. `DataWriter`, `GenericAppenderFactory`), but it does avoid skip computing min/max for fields that don't need them. Please let me know if you are interested, and I'll post a new commit to this PR so that the differences in these two implementations could be compared.

rdblue commented on core/src/main/java/org/apache/iceberg/FieldMetrics.java:
```diff
@@ -30,15 +28,15 @@
-                      ByteBuffer lowerBound,
-                      ByteBuffer upperBound) {
+                      Object lowerBound,
+                      Object upperBound) {
```
If this is going to be the value instead of the serialized value, then the class should be parameterized by the value type: `FieldMetrics<T>` with `public T lowerBound()`.

yyanyy replied:
Yes, sorry I forgot to address that comment after attempting the alternative approach (mentioned in description). I have updated this PR to address this. And sorry for the delay!

rdblue commented:
@yyanyy, can you rebase this? I think merging #1946 caused conflicts.

rdblue commented on core/src/main/java/org/apache/iceberg/FloatFieldMetrics.java:
```diff
@@ -19,16 +19,14 @@
  * exceptions when they are accessed.
  */
-public class FloatFieldMetrics extends FieldMetrics {
+public class FloatFieldMetrics extends FieldMetrics<Object> {
```
Why isn't this `FieldMetrics<Float>`?

yyanyy replied:
Sorry should have double check on this. Although I guess `Number` might fit better since Double writer will also use this class, and `Double` doesn't extend `Float` (although it really shouldn't cause a difference in execution)?

rdblue replied:
Number works for me.

rdblue commented on core/src/main/java/org/apache/iceberg/avro/AvroSchemaUtil.java:
```diff
@@ -408,4 +431,11 @@ private static String sanitize(char character) {
     return "_x" + Integer.toHexString(character).toUpperCase();
   }
+
+  static boolean isMetricSupportedType(Schema.Type type) {
```
How about `supportsMetrics`? I think that would be a bit easier to reason about and it flows in an `if`: `if (supportsMetrics(...)) {...}`

rdblue commented on core/src/main/java/org/apache/iceberg/avro/AvroWithPartnerByStructureVisitor.java:
```diff
@@ -161,4 +186,17 @@ public T map(P sMap, Schema map, T value) {
+  // ---------------------------------- Helpers ---------------------------------------------
+
+  private Deque<String> fieldNames = Lists.newLinkedList();
+  private Deque<Schema> parentSchemas = Lists.newLinkedList();
```
Instead of updating all visitors, why not add extra callbacks like the visitors for Iceberg schemas? I think that supporting `beforeField` and `afterField` would be a better way to handle this than passing the parent and name around. The implementation to get a field's ID from its parent and name seems a bit awkward compared to adding an ID stack in one visitor.

yyanyy replied:
I think I did think of that pattern, and from my notes the main reason I didn't do it is that it won't be as clean as the existing pattern of `before/afterField` in other visitors, as different data structure has different way of retrieving field id information. For example, for fields that are part of the struct, the field id is stored in its own `Schema.Field` so that we can directly pass `field` to `before/afterField` within the for loop when looping through fields; but for map value, the id is stored in the map's schema instead of its own schema, so that `beforeMapValue` should be passed with the parent schema. My thought was that to require different visitors to implement `before/after` with all these different parameters, and to duplicate the various ID retrieval logic among data types in different visitor implementations could be messy to reason about, that keeping them here might actually be cleaner.

rdblue replied:
What about creating a fake `Schema.Field` to pass to the before/after method instead? Another alternative is to pass the field information to the method, like this:

```java
  public beforeField(int fieldId, String name, Schema type);
  public afterField(int fieldId, String name, Schema type);
  public beforeListElement(int elementId, Schema elementType);
  public beforeMapKey(int keyId, Schema keyType);
  public beforeMapValue(int valueId, Schema valueType);
```

I think some variation on this would be better. We want to avoid keeping additional state in all visitors.

rdblue commented on core/src/main/java/org/apache/iceberg/avro/ValueWriters.java:
```diff
@@ -135,145 +150,145 @@ private NullWriter() {
     @Override
     public void write(Void ignored, Encoder encoder) throws IOException {
-      encoder.writeNull();
+      throw new IllegalStateException("[BUG] NullWriter shouldn't be used for writing nulls for Avro");
```
Is this change necessary? Do you think that it would actually cause a problem if it were used?

yyanyy replied:
I think it won't, this was just me trying to make this fail loudly instead of fail silently and to avoid usage confusion in code, I can revert it back if you think it's not necessary.

rdblue replied:
I think we should revert this back so that `writeNull` is correctly called. Nulls shouldn't be written as anything, but the fact that the encoder has a `writeNull` method makes me think that it should be called. We don't know what the encoder might be using it for.

rdblue commented on core/src/main/java/org/apache/iceberg/avro/ValueWriters.java:
```diff
@@ -369,17 +393,42 @@ private OptionWriter(int nullIndex, ValueWriter<T> valueWriter) {
+    private Stream<FieldMetrics> mergeNullCountIntoMetric() {
+      List<FieldMetrics> fieldMetricsFromWriter = valueWriter.metrics().collect(Collectors.toList());
+      Preconditions.checkState(fieldMetricsFromWriter.size() == 1,
+          "Optional field for type % shouldn't vend more than one field metrics", type);
```
Typo: should be `%s` instead of `%`.

rdblue replied:
Also, the use of "vend" is unusual. What about "Optional %s writer should not produce metrics for more than one field"

rdblue replied:
Okay, thinking about this a bit more, I don't think this assumption holds.

There are lots of times when a structure in the middle of a record might be optional. An optional map, for example, would have both key and value metrics. I think that we should only merge null in if we know that this is wrapping a leaf writer, not a map, list, or struct writer.

yyanyy replied:
I think this is currently handled by its caller, that only calls this if `AvroSchemaUtil. isMetricSupportedType` is true, and `AvroSchemaUtil. isMetricSupportedType` only returns true for primitive types?

rdblue replied:
Yes, you're right that it is handled there. I missed that originally. I commented on how to avoid passing the type in, but I agree that you do handle it.

rdblue commented on core/src/main/java/org/apache/iceberg/avro/ValueWriters.java:
```diff
@@ -358,8 +380,10 @@ public void write(BigDecimal decimal, Encoder encoder) throws IOException {
+    private long nullValueCount;
 
-    private OptionWriter(int nullIndex, ValueWriter<T> valueWriter) {
+    private OptionWriter(int nullIndex, ValueWriter<T> valueWriter, Schema.Type type) {
```
It looks like the only place that `type` is used is in an error message. I'd prefer not to change the signature here just to print a type.

yyanyy replied:
Oh I think this `type` is also used for checking `isMetricSupportedType/supportsMetrics` so we will still need it.

rdblue replied:
I think that we should not pass the type here. Instead, let's have a `MetricsOptionWriter` and a non-metrics version and choose which one to use ahead of time. That way, we don't keep a count that won't be used. There's no need to check whether the null count should be used when `metrics` is called. That can be determined ahead of time.

If we do that, then there is no need to pass the type here or into the `option` method. We can either add a boolean (`collectMetrics`) or have a separate factory method. I think that would be a bit cleaner.

yyanyy reviewed (commented):
Thank you @rdblue for reviewing this PR! I have rebased the changes and responded/addressed the feedback in this PR and #1946 except for the one that I forgot to include `rowWriter` as part of the `metrics()` for positional delete writer, which I'll do in a separate PR since I wanted to have both code and test coverage in the same PR for that.

rdblue commented on core/src/main/java/org/apache/iceberg/avro/AvroFileAppender.java:
```diff
@@ -95,10 +95,10 @@ public void close() throws IOException {
       CodecFactory codec, Map<String, String> metadata) throws IOException {
     DataFileWriter<D> writer = new DataFileWriter<>(
-        (DatumWriter<D>) metricsAwareDatumWriter);
+        (DatumWriter<D>) datumWriter);
```
Nit: no need for the newline any more.

rdblue commented on core/src/main/java/org/apache/iceberg/avro/AvroSchemaUtil.java:
```diff
@@ -244,6 +245,28 @@ static boolean hasProperty(Schema schema, String propertyName) {
     return schema.getObjectProp(propertyName) != null;
   }
 
+  public static int fieldId(Schema currentSchema, Schema parentSchema, Supplier<String> fieldNameGetter) {
```
I'd like to avoid needing this method because of the strange input types. I think those show that the way we're traversing the schema isn't quite right, which is why this is complicated.

rdblue commented on core/src/main/java/org/apache/iceberg/avro/ValueWriters.java:
```diff
@@ -135,145 +150,145 @@ private NullWriter() {
-    private static final BooleanWriter INSTANCE = new BooleanWriter();
+    @Override
+    public Stream<FieldMetrics> metrics() {
+      throw new IllegalStateException("[BUG] NullWriter shouldn't be used for writing nulls for Avro");
```
Let's default this to `Stream.empty()` so that it won't fail.

rdblue commented on core/src/main/java/org/apache/iceberg/avro/ValueWriters.java:
```diff
@@ -135,145 +150,145 @@ private NullWriter() {
-    private ByteToIntegerWriter() {
+  private static class ByteToIntegerWriter extends MetricsAwareTransformWriter<Byte, Integer> {
+    private ByteToIntegerWriter(int id) {
+      super(id, Integer::compareTo, Byte::intValue);
```
This change is will lead to unnecessary boxing of the converted value because it needs to pass an Object to `writeVal`. I'd rather duplicate a little bit more code than have the object creation in a tight loop. Can you update the byte and short writers to avoid this?

rdblue replied:
Actually, I think it's fine this way.

rdblue commented on core/src/main/java/org/apache/iceberg/avro/ValueWriters.java:
```diff
@@ -135,145 +150,145 @@ private NullWriter() {
-  private static class Utf8Writer implements ValueWriter<Utf8> {
-    private static final Utf8Writer INSTANCE = new Utf8Writer();
+    @Override
+    protected void writeVal(CharSequence s, Encoder encoder) throws IOException {
```
Is this called? Seems like it wouldn't be since `write` is overridden.

rdblue replied:
Nevermind, `write` delegates to `super.write`.

rdblue commented on core/src/main/java/org/apache/iceberg/avro/ValueWriters.java:
```diff
@@ -492,4 +561,153 @@ protected Object get(IndexedRecord struct, int pos) {
+   * @param <T2> Type after transformation
+   */
+  @SuppressWarnings("checkstyle:VisibilityModifier")
+  public abstract static class MetricsAwareTransformWriter<T1, T2> implements ValueWriter<T1> {
```
We typically use `S` and `T` for input and output types in other transform classes.

rdblue commented on core/src/main/java/org/apache/iceberg/avro/ValueWriters.java:
```diff
@@ -492,4 +561,153 @@ protected Object get(IndexedRecord struct, int pos) {
+    @Override
+    public void write(T1 datum, Encoder encoder) throws IOException {
+      valueCount++;
+      if (datum == null) {
```
Is this necessary? Before, all of the type-specific writers assumed that the input value was non-null because null isn't allowed unless the type is optional. If it is optional, the writer will be wrapped in an option writer, so there is no need to handle null.

I think this should similarly assume that the option writer tracks null values and that all values will be non-null. That simplifies this class quite a bit.

rdblue commented on core/src/main/java/org/apache/iceberg/avro/ValueWriters.java:
```diff
@@ -492,4 +561,153 @@ protected Object get(IndexedRecord struct, int pos) {
+          min = transformedDatum;
+        }
+        writeVal(transformedDatum, encoder);
+
```
Nit: no newline after `if` blocks and unnecessary newline before closing curly.

rdblue replied:
I see this in a couple other places, too.

rdblue commented on core/src/main/java/org/apache/iceberg/avro/ValueWriters.java:
```diff
@@ -492,4 +561,153 @@ protected Object get(IndexedRecord struct, int pos) {
   }
+
+  private abstract static class FloatingPointWriter<T extends Comparable<T>>
+      extends ComparableWriter<T> {
```
I don't think this needs to extend `ComparableWriter` since it calls `compareTo` directly.

yyanyy replied:
I think this was to mainly save an argument for constructor, but I can pass `null` and add a comment.

rdblue commented on core/src/main/java/org/apache/iceberg/avro/ValueWriters.java:
```diff
@@ -492,4 +561,153 @@ protected Object get(IndexedRecord struct, int pos) {
+        if (min == null || comparator.compare(transformedDatum, min) < 0) {
+          min = transformedDatum;
+        }
+        writeVal(transformedDatum, encoder);
```
After removing the null handling from this method, this is the main block of code provided by this class. I think that the class structure here tries too hard to avoid duplication of these 8 lines and ends up causing an unnecessary performance slow down.

Most of the subclasses don't actually provide a transform, so it doesn't make a lot of sense to provide one. In addition, the cases that do provide a transform only produce `int` and `long` values. Using this inheritance structure, all writers will apply a function and the `int` and `long` cases will box the result of that function call.

I think separating this into `StoredAsIntwriter`, `StoredAsLongWriter`, and `MetricsAwareWriter` would duplicate these lines, but avoid the function call and boxing for a lot of cases. It would also avoid needing to implement `writeVal(Integer, Encoder)` in every subclass. Instead, you'd need to add `int convert(T)` to convert to the int value before writing to the encoder directly in `write`.

Could you update to use that structure?

yyanyy replied:
Yes, this makes the code much cleaner than before, thank you for the great suggestion!

rdblue commented on core/src/main/java/org/apache/iceberg/avro/ValueWriters.java:
```diff
@@ -492,4 +561,153 @@ protected Object get(IndexedRecord struct, int pos) {
+    }
+  }
+
+  public abstract static class MetricsAwareStringWriter<T extends Comparable<T>> extends ComparableWriter<T> {
```
Is this only used for Spark's UTF8String?

yyanyy replied:
Flink's StringData writer also extends it.

yyanyy commented:
Thank you @rdblue for reviewing this PR again! I think I have addressed all the feedback. Since the visitor change is non-trivial and the original change didn't contain tests, in order to ensure the new code with before/afterField pattern doesn't impact normal avro writes, I added all the necessary tests to this PR (which means that #1935 will be done if the current changes are merged), but please feel free to suggest removing some classes for an easier review.

One thing I'd like to mention is a behavior difference among different file types in handling optional struct field with required and optional sub fields. I think it might be fine for now since I think we don't use those statistics, but just want to capture the current status here:

for example, if the schema is `someField, optional struct<required int1, optional int2>`, and there are 50 records with this schema in total, within which 6 records have null struct, and 3 records have struct but null int2, then each file format will produce the following count info:

|         | valueCount | nullCount |
|---------|------------|-----------|
| Parquet | int1: 50, int2: 50       | int1:  6,  int2: 9    |
| Orc     |int1: 44, int2: 50            | int1: 0, int2: 9          |
| Avro    |int1: 44, int2: 44            |int1: 0, int2: 3          |

rdblue commented on core/src/main/java/org/apache/iceberg/FloatFieldMetrics.java:
```diff
@@ -19,16 +19,14 @@
  * exceptions when they are accessed.
  */
-public class FloatFieldMetrics extends FieldMetrics {
+public class FloatFieldMetrics extends FieldMetrics<Number> {
```
After reviewing #2464, I think I understand why `Number` is used here instead of `Float` or `Double`, but I think it would be better to make each `FieldMetrics` class specific to a value type. This will probably be done when #2464 is merged and this is rebased for float, but in the mean time you may want to update this for any other type metrics you're introducing in this PR.

rdblue commented on core/src/main/java/org/apache/iceberg/avro/AvroFileAppender.java:
```diff
@@ -95,10 +95,9 @@ public void close() throws IOException {
       CodecFactory codec, Map<String, String> metadata) throws IOException {
-    DataFileWriter<D> writer = new DataFileWriter<>(
-        (DatumWriter<D>) metricsAwareDatumWriter);
+    DataFileWriter<D> writer = new DataFileWriter<>((DatumWriter<D>) datumWriter);
```
Is this rename needed? While I support cleaning up names, I would generally opt to leave these as-is to have smaller commits that are less likely to cause conflicts.

rdblue commented on core/src/main/java/org/apache/iceberg/avro/AvroMetrics.java:
```diff
@@ -31,7 +42,112 @@ private AvroMetrics() {
-    // TODO will populate in following PRs if datum writer is a MetricsAwareDatumWriter
-    return new Metrics(numRecords, null, null, null);
+    if (!(datumWriter instanceof MetricsAwareDatumWriter)) {
+      return new Metrics(numRecords, null, null, null, null);
```
is `numRecords` correct? What if this is for a field nested in a map or list?

rdblue commented on core/src/main/java/org/apache/iceberg/avro/AvroMetrics.java:
```diff
@@ -31,7 +42,112 @@ private AvroMetrics() {
+
+    metricsAwareDatumWriter.metrics().forEach(metrics -> {
+      String columnName = schema.findColumnName(metrics.id());
+      MetricsModes.MetricsMode metricsMode = metricsConfig.columnMode(columnName);
```
Should we add a method to look up metrics mode by ID?

rdblue commented on core/src/main/java/org/apache/iceberg/avro/AvroMetrics.java:
```diff
@@ -31,7 +42,112 @@ private AvroMetrics() {
+      switch (type.typeId()) {
+        case STRING:
+          lowerBound = UnicodeUtil.truncateStringMin(
+              Literal.of((CharSequence) metrics.lowerBound()), truncateLength).value();
```
Why not use `UnicodeUtil.truncateStringMin(CharSequence, int)` instead of the `Literal` method?

rdblue replied:
Ah, I see that the `min` and `max` methods take literals. I think it would be better to refactor those methods to expose non-Literal implementations instead of creating the literals and unwrapping the result.

rdblue commented on core/src/main/java/org/apache/iceberg/avro/AvroMetrics.java:
```diff
@@ -31,7 +42,112 @@ private AvroMetrics() {
+      }
+
+      updateLowerBound(metrics, type, metricsMode).ifPresent(lowerBound -> lowerBounds.put(metrics.id(), lowerBound));
+      updateUpperBound(metrics, type, metricsMode).ifPresent(upperBound -> upperBounds.put(metrics.id(), upperBound));
```
I don't quite understand the decision to return an option instead of just passing `lowerBounds` or `upperBounds` into the update method and having the `put` happen there. Wouldn't it be simpler if the put happened at the end of `updateLowerBound` or `updateUpperBound`?

rdblue commented on core/src/main/java/org/apache/iceberg/avro/AvroMetrics.java:
```diff
@@ -31,7 +42,112 @@ private AvroMetrics() {
+      metricsConfig = inputMetricsConfig;
+    }
+
+    Map<Integer, Long> valueCounts = new HashMap<>();
```
Nit: we typically prefer `Maps.newHashMap()`, which will add null checking, and no put value should be null here.

rdblue commented on core/src/main/java/org/apache/iceberg/avro/AvroMetrics.java:
```diff
@@ -31,7 +42,112 @@ private AvroMetrics() {
+    Map<Integer, Long> nullValueCounts = new HashMap<>();
+    Map<Integer, Long> nanValueCounts = new HashMap<>();
+    Map<Integer, ByteBuffer> lowerBounds = new HashMap<>();
+    Map<Integer, ByteBuffer> upperBounds = new HashMap<>();
```
One thing to consider is that quite a bit of this method and the helper methods below is generic and could be written for `Stream<FieldMetrics>`. I don't think it needs to happen right now, but I think it would be good to have this separated into the Avro-specific part and a metrics part that lives in `MetricsUtil`.

rdblue commented on core/src/main/java/org/apache/iceberg/avro/AvroSchemaVisitor.java:
```diff
@@ -59,13 +63,22 @@
         if (schema.getLogicalType() instanceof LogicalMap) {
-          return visitor.array(schema, visit(schema.getElementType(), visitor));
+          T result = visit(schema.getElementType(), visitor);
+          return visitor.array(schema, result);
```
It doesn't look like this change is needed?

rdblue commented on core/src/main/java/org/apache/iceberg/avro/AvroWithPartnerByStructureVisitor.java:
```diff
@@ -51,7 +51,12 @@
             "Invalid map: %s is not a string", keyType);
-        return visitor.map(partner, schema, visit(visitor.mapValueType(partner), schema.getValueType(), visitor));
+
+        visitor.beforeMapValue("value", schema.getValueType(), schema);
```
Should this make an artificial call to the map key callbacks using a generic String schema as well?

rdblue commented on core/src/main/java/org/apache/iceberg/avro/AvroWriterBuilderFieldIdUtil.java:
```diff
@@ -0,0 +1,58 @@
+ * This util class helps Avro DatumWriter builders to retrieve the correct field Id when building Avro DatumWriters
+ * with visitor pattern.
+ */
+public class AvroWriterBuilderFieldIdUtil {
```
Minor: I think it would be a bit shorter to make this an abstract class that can be extended to get the current field ID via a protected method, but this is a good solution already.

rdblue commented on core/src/main/java/org/apache/iceberg/avro/ValueWriters.java:
```diff
@@ -492,4 +576,215 @@ protected Object get(IndexedRecord struct, int pos) {
+  public abstract static class StoredAsIntWriter<T> implements ValueWriter<T> {
+    protected final int id;
+    protected long valueCount;
+    protected Integer max;
```
Because this is only used for non-null values, I think you can simplify the min and max updates by removing the null check. Just use `protected int min = Integer.MAX_VALUE` here and remove the null check. Then when creating `FieldMetrics`, you can check whether `valueCount` is non-zero to know whether `min` and `max` are valid.

rdblue commented on core/src/main/java/org/apache/iceberg/avro/ValueWriters.java:
```diff
@@ -492,4 +576,215 @@ protected Object get(IndexedRecord struct, int pos) {
     }
   }
+
+  private abstract static class FloatingPointWriter<T extends Comparable<T>>
```
After #2464, would this be needed? Or would you just use the floating point classes from that PR?

rdblue commented on core/src/main/java/org/apache/iceberg/avro/ValueWriters.java:
```diff
@@ -137,143 +156,143 @@ private NullWriter() {
-    private static final LongWriter INSTANCE = new LongWriter();
-
-    private LongWriter() {
+  private static class LongWriter extends ComparableWriter<Long> {
```
I would expect `LongWriter` to extend `StoredAsLongWriter` instead of `ComparableWriter`. Same for `IntegerWriter`. Could you update those?

rdblue commented on core/src/main/java/org/apache/iceberg/avro/ValueWriter.java:
```diff
@@ -27,7 +27,5 @@
-  default Stream<FieldMetrics> metrics() {
-    return Stream.empty(); // TODO will populate in following PRs
-  }
+  Stream<FieldMetrics> metrics();
```
`FieldMetrics` is parameterized, but this is a bare reference. Could you update it? I think it should be `FieldMetrics<?>` since the metrics are not necessarily for the written value type, `D`.

rdblue commented on core/src/main/java/org/apache/iceberg/avro/ValueWriters.java:
```diff
@@ -492,4 +576,215 @@ protected Object get(IndexedRecord struct, int pos) {
+      writeVal(datum, encoder);
+    }
+
+    protected abstract void writeVal(T datum, Encoder encoder) throws IOException;
```
I would probably name this `encode` rather than `writeVal` so that there is less confusion with the `write` method.

rdblue commented on core/src/main/java/org/apache/iceberg/avro/ValueWriters.java:
```diff
@@ -492,4 +576,215 @@ protected Object get(IndexedRecord struct, int pos) {
+    protected abstract void writeVal(T datum, Encoder encoder) throws IOException;
+
+    @Override
+    public Stream<FieldMetrics> metrics() {
```
I'd like to fix all of the references that don't parameterize `FieldMetrics`. I think they should be `FieldMetrics<?>`.

rdblue commented on core/src/test/java/org/apache/iceberg/TestMetrics.java:
```diff
@@ -274,7 +282,11 @@ public void testMetricsForNestedStructFields() throws IOException {
         ByteBuffer.wrap("A".getBytes()), ByteBuffer.wrap("A".getBytes()), metrics);
     assertCounts(7, 1L, 0L, 1L, metrics);
-    assertBounds(7, DoubleType.get(), Double.NaN, Double.NaN, metrics);
+    if (fileFormat() == FileFormat.AVRO) {
```
Is this needed if #2464 goes in first?

rdblue commented on core/src/test/java/org/apache/iceberg/TestMetrics.java:
```diff
@@ -643,4 +747,10 @@ protected void assertCounts(int fieldId, Long valueCount, Long nullValueCount, L
+  private void assertNonNullColumnSizes(Metrics metrics) {
+    if (fileFormat() != FileFormat.AVRO) {
+      Assert.assertTrue(metrics.columnSizes().values().stream().allMatch(Objects::nonNull));
+    }
```
Can you add an assertion for Avro? I think that the column sizes map should be null, is that correct?

rdblue commented on data/src/test/java/org/apache/iceberg/avro/TestGenericAvroMetrics.java:
```diff
@@ -0,0 +1,50 @@
+public class TestGenericAvroMetrics extends TestAvroMetrics {
+
+  protected Metrics getMetrics(Schema schema, OutputFile file, Map<String, String> properties,
+                                        MetricsConfig metricsConfig, Record... records) throws IOException {
```
Nit: indentation is off.

rdblue commented on core/src/main/java/org/apache/iceberg/avro/ValueWriters.java:
```diff
@@ -45,68 +55,72 @@ private ValueWriters() {
-  public static <T> ValueWriter<T> option(int nullIndex, ValueWriter<T> writer) {
-    return new OptionWriter<>(nullIndex, writer);
+  public static <T> ValueWriter<T> option(int nullIndex, ValueWriter<T> writer, Schema.Type type) {
+    if (AvroSchemaUtil.supportsMetrics(type)) {
```
Rather than introducing `supportsMetrics`, why not just check the value writer to see whether it is a `MetricsAwareWriter`? I know that not all of the writers extend that class, but you could either introduce a `MetricsWriter` interface to signal that the inner writer supports metrics that all of the implementations extend, or maybe you could alter the hierarchy a little so that `StoredAsIntWriter` and `StoredAsLongWriter` actually do extend `MetricsAwareWriter`. Then you wouldn't need changes to `AvroSchemaUtil` or so many changes to option handling.

rdblue commented on flink/src/main/java/org/apache/iceberg/flink/data/FlinkValueWriters.java:
```diff
@@ -74,52 +78,60 @@ private FlinkValueWriters() {
+
+    @Override
+    public Stream<FieldMetrics> metrics() {
+      return metrics(DecimalData::toBigDecimal);
```
Does this return a Java `BigDecimal`? I know we've had problems with this method in Spark because it actually produces a Scala `BigDecimal`.

rdblue commented on spark/src/main/java/org/apache/iceberg/spark/data/SparkValueWriters.java:
```diff
@@ -72,61 +71,41 @@ private SparkValueWriters() {
+
+    @Override
+    public Stream<FieldMetrics> metrics() {
+      return metrics(Decimal::toJavaBigDecimal);
```
Good to see you got the right one.

rdblue commented:
Any update on this one, @yyanyy?

yyanyy commented:
> Any update on this one, @yyanyy?

Apologies, didn't find a chance to update this. I'll make sure to allocate time to address the comments in the coming two weeks!

rdblue commented:
Thanks, @yyanyy! No rush, I just wanted to check in.
